src/modules/Game.class.js

```diff
@@ -1,68 +1,271 @@
-   *
-   * @param {number[][]} initialState
-   * The initial state of the board.
-   * @default
-   * [[0, 0, 0, 0],
-   *  [0, 0, 0, 0],
-   *  [0, 0, 0, 0],
-   *  [0, 0, 0, 0]]
-   *
-   * If passed, the board will be initialized with the provided
-   * initial state.
-   */
-  constructor(initialState) {
-    // eslint-disable-next-line no-console
-    console.log(initialState);
-  }
-
-  moveLeft() {}
-  moveRight() {}
-  moveUp() {}
-  moveDown() {}
-
-  /**
-   * @returns {number}
-   */
-  getScore() {}
-
-  /**
-   * @returns {number[][]}
-   */
-  getState() {}
-
-  /**
-   * Returns the current game status.
-   *
-   * @returns {string} One of: 'idle', 'playing', 'win', 'lose'
-   *
-   * `idle` - the game has not started yet (the initial state);
-   * `playing` - the game is in progress;
-   * `win` - the game is won;
-   * `lose` - the game is lost
-   */
-  getStatus() {}
-
-  /**
-   * Starts the game.
-   */
-  start() {}
-
-  /**
-   * Resets the game.
-   */
-  restart() {}
-
-  // Add your own methods here
+  static Status = {
+    idle: 'idle',
+    playing: 'playing',
+    win: 'win',
+    lose: 'lose',
+  };
+
+  constructor(
+    initialState = [
+      [0, 0, 0, 0],
+      [0, 0, 0, 0],
+      [0, 0, 0, 0],
+      [0, 0, 0, 0],
+    ],
+  ) {
+    this.initialState = initialState;
+    this.state = this.initialState.map((row) => [...row]);
+    this.status = Game.Status.idle;
+    this.score = 0;
+  }
+
+  moveLeft() {
+    if (this.getStatus() !== Game.Status.playing) {
+      return;
+    }
+
+    const newState = this.state.map((row) => this.move(row));
+
+    if (JSON.stringify(this.getState()) !== JSON.stringify(newState)) {
+      this.updateGameState(newState);
+      this.insertTiles();
+    }
+  }
+
+  moveRight() {
+    if (this.getStatus() !== Game.Status.playing) {
+      return;
+    }
+
+    const reversedState = this.state.map((row) => [...row].reverse());
+
+    const newState = reversedState.map((row) => {
+      return this.move(row).reverse();
+    });
+
+    if (JSON.stringify(this.getState()) !== JSON.stringify(newState)) {
+      this.updateGameState(newState);
+      this.insertTiles();
+    }
+  }
+
+  moveUp() {
+    if (this.getStatus() !== Game.Status.playing) {
+      return;
+    }
+
+    const rotateState = this.rotateLeft(this.getState());
+
+    const newState = rotateState.map((row) => this.move(row));
+
+    const unRotateState = this.rotateRight(newState);
+
+    if (JSON.stringify(this.getState()) !== JSON.stringify(unRotateState)) {
+      this.updateGameState(unRotateState);
+      this.insertTiles();
+    }
+  }
+
+  moveDown() {
+    if (this.getStatus() !== Game.Status.playing) {
+      return;
+    }
+
+    const rotateState = this.rotateRight(this.getState());
+
+    const newState = rotateState.map((row) => this.move(row));
+
+    const unRotateState = this.rotateLeft(newState);
+
+    if (JSON.stringify(this.getState()) !== JSON.stringify(unRotateState)) {
```

Comment:
Avoid using `JSON.stringify` for state comparison to improve performance, as mentioned in previous comments.